File: ruby/sk3/task.rb

module SK3

  class Task
    attr_reader :name, :batch_time, :batch_size, :buffer, :low_water_mark
    attr_accessor :machine

    def initialize(event_queue, _name,
                   _batch_time, _batch_size, _buffer = 0)
      @event_queue = event_queue
      @name        = _name
      @batch_time  = _batch_time
      @batch_size  = _batch_size
      @buffer      = _buffer
      @prereqs = {}
      @demands = []
    end

    def init_sim
      fail "Task #{@name} isn't assigned to a machine" unless @machine
      if @buffer < 0
        @start_time = @event_queue.curtime
        @machine.enqueue(self)
      end
      reset_low_water_mark!
    end

    def reset_low_water_mark!
      @low_water_mark = @buffer
    end

    def add_prereq(task, quantity)
      if @prereqs[task]
        fail "Duplicate requirement: #{name} already requires #{task.name}"
      end
      @prereqs[task] = quantity
    end

    def start_batch
      @prereqs.each do |task,quantity|
        task.take_from_buffer(quantity, @demands.first)
      end
    end

    def finish_batch
      produced = @batch_size
      self.buffer += produced
      while (produced > 0) && @demands.first
        produced = @demands.first.fulfilling(self, produced)
        if produced
          @demands.shift
        else
          produced = 0
        end
      end
      if @buffer < 0
        @machine.enqueue(self)
      else
        log_filled_buffer
      end
    end

    def take_from_buffer(quantity, demand)
      if (@buffer >= 0) && (@buffer < quantity)
        @start_time = @event_queue.curtime
      end
      self.buffer -= quantity
      @low_water_mark = @buffer if @buffer < @low_water_mark
      @machine.enqueue(self) if should_enqueue?
      return if @buffer >= 0
      if demand
        if quantity < -@buffer
          demand.taking(self, -@buffer)
        else
          demand.taking(self, quantity)
        end
        @demands.push demand unless @demands.include? demand
      end
    end

    # hook for subclass
    def should_enqueue?
      @buffer < 0
    end

    def hash
      @name.hash
    end

    protected

    def buffer=(quantity)
      @buffer = quantity
      @event_queue.log(:deficit, self, 0)
    end

    private

    def log_filled_buffer
      length = @event_queue.curtime - @start_time
      @event_queue.log(:buffer_full, self, length)
    end

  end

end

File: cpp/src/task.cpp
#include <algorithm>
#include "task.hpp"
#include "machine.hpp"
#include "demand.hpp"
#include "event_queue.hpp"

namespace SK3 {

using namespace std;

namespace {

template<class T> shared_ptr<Task> factory(const Config::Task &cfg,
    shared_ptr<EventQueue> eventQ) {
  return make_shared<T>(eventQ, cfg.name,
      to_internal_time(cfg.batch_time),
      to_internal_quantity(cfg.batch_size),
      to_internal_quantity(cfg.deficit));
}

static map<string, task_factory_t> factories {
  { "Task", factory<Task> }
};

} // anon namespace


task_factory_t Task::factoryFor(const string &variant) {
  return factories.at(variant);
}

Task::Task(shared_ptr<EventQueue> _eventQ, const string &_name,
    const Time _batch_time, const Quantity _batch_size,
    const Quantity init_buffer):
  taskName(_name),
  batch_time(_batch_time),
  batch_size(_batch_size),
  eventQ(_eventQ),
  taskBuffer(init_buffer),
  low_water_mark(init_buffer) { }

void Task::init_sim() {
  if (taskBuffer < 0) {
    startTime = eventQ->now();
    machine.lock()->enqueue(shared_from_this());
  }
  resetLowWaterMark();
}

const string &Task::name() const { return taskName; }
const Quantity Task::buffer() const { return taskBuffer; }

void Task::take_from_buffer(const Quantity quantity,
    shared_ptr<DemandOrder> order) {
  if ((taskBuffer >= 0) && (taskBuffer < quantity)) {
    startTime = eventQ->now();
  }
  taskBuffer -= quantity;
  if (taskBuffer < low_water_mark) {
    low_water_mark = taskBuffer;
  }
  if (should_enqueue()) {
    machine.lock()->enqueue(shared_from_this());
  }
  if (taskBuffer >= 0) return;
  //FIXME: max seems wrong; I think it should be min?
  order->taking(shared_from_this(), max(quantity, -taskBuffer));
  if (find(orders.begin(), orders.end(), order) == orders.end()) {
    orders.push_back(order);
  }
}

void Task::startBatch() {
  shared_ptr<DemandOrder> serving = orders.front();
  for_each(suppliers.begin(), suppliers.end(),
      [serving](Supplier &supplier) {
        shared_ptr<Task> task = supplier.task.lock();
        task->take_from_buffer(supplier.quantity, serving);
      });
}

typedef EventQueue::LogRecord LogRecord;
typedef LogRecord::Type LogType;

void Task::finishBatch() {
  Quantity produced = batch_size;
  taskBuffer += produced;
  while ((produced > 0) && !orders.empty()) {
    pair<bool, Quantity> fulfilled =
      orders.front()->fulfilling(shared_from_this(), produced);
    if (fulfilled.first) {
      orders.pop_front();
    }
    produced = fulfilled.second;
  }
  if (taskBuffer < 0) {
    machine.lock()->enqueue(shared_from_this());
  } else {
    Time duration = eventQ->now() - startTime;
    eventQ->log(
        LogRecord(LogType::LOG_BUFFER_FULL, duration),
        *this);
  }
}

Quantity Task::resetLowWaterMark() {
  Quantity low = low_water_mark;
  low_water_mark = taskBuffer;
  return low;
}

bool Task::should_enqueue() {
  return taskBuffer < 0;
  return false;
}

}

